// @adonisjs/mail/build/src/Fake/index.js
"use strict";
/*
 * @adonisjs/mail
 *
 * (c) AdonisJS
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FakeMailManager = void 0;
class FakeMailManager {
    constructor() {
        this.fakedMailers = new Map();
    }
    /**
     * Returns the faked mailer instance
     */
    use(mailer) {
        return this.fakedMailers.get(mailer);
    }
    /**
     * Restore mailer fake
     */
    restore(mailer) {
        const mailerInstance = this.fakedMailers.get(mailer);
        if (mailerInstance) {
            mailerInstance.close();
            this.fakedMailers.delete(mailer);
        }
    }
    /**
     * Find if a mailer is faked
     */
    isFaked(mailer) {
        return this.fakedMailers.has(mailer);
    }
    /**
     * Find if an email exists
     */
    exists(messageOrCallback) {
        return !!this.find(messageOrCallback);
    }
    /**
     * Find an email
     */
    find(messageOrCallback) {
        for (let [, mailer] of this.fakedMailers) {
            const message = mailer.driver.find(messageOrCallback);
            if (message) {
                return message;
            }
        }
        return null;
    }
    /**
     * Filter emails
     */
    filter(messageOrCallback) {
        let messages = [];
        for (let [, mailer] of this.fakedMailers) {
            messages = messages.concat(mailer.driver.filter(messageOrCallback));
        }
        return messages;
    }
}
exports.FakeMailManager = FakeMailManager;
